src/MediaWiki.ts

```diff
@@ -3,68 +3,143 @@ import * as logger from './Logger.js'
-  public readonly modulePath: string
-  public readonly webUrl: URL
-  public readonly apiUrl: URL
-  public readonly veApiUrl: URL
-  public readonly restApiUrl: URL
-  public readonly mobileRestApiUrl: URL
-  public readonly desktopRestApiUrl: URL
-  public readonly getCategories: boolean
-  public readonly namespaces: MWNamespaces = {}
-  public readonly namespacesToMirror: string[] = []
-
-  private readonly wikiPath: string
-  private readonly username: string
-  private readonly password: string
-  private readonly apiPath: string
-  private readonly domain: string
+  public _base: string
+  public baseUrl: URL
+  public getCategories: boolean
+  public namespaces: MWNamespaces = {}
+  public namespacesToMirror: string[] = []
+  public mwArticleId: string
+
+  #wikiPath: string
+  #restApiPath: string
+  #username: string
+  #password: string
+  #apiPath: string
+  #domain: string
   private apiUrlDirector: ApiURLDirector
+  private wikimediaDesktopUrlDirector: DesktopURLDirector
+  private visualEditorURLDirector: VisualEditorURLDirector
 
-  constructor(config: MWConfig) {
-    this.domain = config.domain || ''
-    this.username = config.username
-    this.password = config.password
-    this.getCategories = config.getCategories
+  public visualEditorApiUrl: URL
+  public apiUrl: URL
+  public modulePath: string // only for reading
+  public _modulePathOpt: string // only for whiting to generate modulePath
+  public webUrl: URL
+  public desktopRestApiUrl: URL
 
-    this.baseUrl = basicURLDirector.buildMediawikiBaseURL(config.base)
+  #hasWikimediaDesktopRestApi: boolean | null
+  #hasVisualEditorApi: boolean | null
 
-    this.apiPath = config.apiPath ?? 'w/api.php'
-    this.wikiPath = config.wikiPath ?? DEFAULT_WIKI_PATH
+  set username(value: string) {
+    this.#username = value
+  }
 
-    const baseUrlDirector = new BaseURLDirector(this.baseUrl.href)
+  set password(value: string) {
+    this.#password = value
+  }
 
-    this.webUrl = baseUrlDirector.buildURL(this.wikiPath)
-    this.apiUrl = baseUrlDirector.buildURL(this.apiPath)
+  set apiPath(value: string) {
+    this.#apiPath = value
+  }
 
-    this.apiUrlDirector = new ApiURLDirector(this.apiUrl.href)
+  set restApiPath(value: string) {
+    this.#restApiPath = value
+  }
+
+  set domain(value: string) {
+    this.#domain = value
+  }
 
-    this.veApiUrl = this.apiUrlDirector.buildVisualEditorURL()
+  set wikiPath(value: string) {
+    this.#wikiPath = value
+  }
 
-    this.restApiUrl = baseUrlDirector.buildRestApiURL(config.restApiPath)
-    this.desktopRestApiUrl = baseUrlDirector.buildDesktopRestApiURL(config.restApiPath)
+  set base(value: string) {
+    this.baseUrl = basicURLDirector.buildMediawikiBaseURL(value)
+    this.initMWApis()
+  }
 
-    this.modulePath = baseUrlDirector.buildModuleURL(config.modulePath)
+  set modulePathOpt(value: string) {
+    this._modulePathOpt = value
+  }
+
+  private initializeMediaWikiDefaults(): void {
```

Comment:
I don't think I understand your concern about Mediawiki properties initializing and resetting.
The `initializeMediaWikiDefaults()` method provides default Mediawiki properties that have to be reset when the execution script in `src/mwoffliner.lib.ts` is complete. We init this list of props inside constructor when create Mediawiki singleton, set credentials, capabilities, and base URL. And then we need to reset all of this because singleton is used around the application.

Reply:
Makes sense